include/aparse/abstract_core_parser.hpp

```diff
@@ -0,0 +1,83 @@
+// Copyright: 2015 Mohit Saini
+// Author: Mohit Saini (mohitsaini1196@gmail.com)
+
+#ifndef _APARSE_ABSTRACT_CORE_PARSER_HPP_
+#define _APARSE_ABSTRACT_CORE_PARSER_HPP_
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <quick/utility.hpp>
+#include <quick/debug_stream_decl.hpp>
+
+#include "aparse/error.hpp"
+
+namespace aparse {
+
+// Node of ParseTree.
+struct CoreParseNode {
+  CoreParseNode() {}
+  explicit CoreParseNode(int label): label(label) {}
+  CoreParseNode(int label, int start): label(label), start(start) {}
+  CoreParseNode(int label, const pair<int, int>& range): label(label),
+                                                         start(range.first),
+                                                         end(range.second) {}
+  CoreParseNode(int label,
+                const pair<int, int>& range,
+                const vector<CoreParseNode>& children): label(label),
+                                                        start(range.first),
+                                                        end(range.second),
+                                                        children(children) {}
+  CoreParseNode(int label,
+                const pair<int, int>& range,
+                vector<CoreParseNode>&& children)
+       : label(label),
+         start(range.first),
+         end(range.second),
+         children(std::move(children)) {}
+  bool operator==(const CoreParseNode& rhs) const;
+  void DebugStream(qk::DebugStream& ds) const;
+
+  int label = 0;
+  int start = 0, end = 0;  // start: inclusive, end: exclusive;
+  vector<CoreParseNode> children;
+};
+
+
+class AbstractCoreParser {
```

Comment:
Internally, we have followed the practice of ensuring that with each class we add in documentation explaining the following:
* What is the purpose of the class ?
* What is a sample usage of the class (preferably with an example) ?

Reply:
I have updated the purpose of each classes.
Sample usage can be found in corresponding test for each class.